compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirContractChecker.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.contracts.FirResolvedContractDescription
+import org.jetbrains.kotlin.fir.declarations.FirContractDescriptionOwner
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.declarations.FirProperty
+import org.jetbrains.kotlin.fir.expressions.FirLoop
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.expressions.FirWhenExpression
+
+private const val UNEXPECTED_CONSTRUCTION = "Unexpected construction in contract description"
+
+object FirContractChecker : FirFunctionChecker() {
+    override fun check(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration !is FirContractDescriptionOwner ||
+            declaration.contractDescription !is FirResolvedContractDescription
+        ) {
+            return
+        }
+
+        for (statement in (declaration.contractDescription as FirResolvedContractDescription).unresolvedEffects) {
+            if (statement.source == null || statement.source!!.kind is FirFakeSourceElementKind) continue
+
+            if (statement.isNotAllowedInContract) {
+                reporter.report(FirErrors.ERROR_IN_CONTRACT_DESCRIPTION.on(statement.source!!, UNEXPECTED_CONSTRUCTION), context)
+            }
+        }
+    }
+
+    private val FirStatement.isNotAllowedInContract: Boolean
```

Comment:
How did you come up with this condition? If this is the equivalent condition in FE1.0 (https://github.com/JetBrains/kotlin/blob/9c1b68f8394faf4c4742dc7788e5b0365d8d4a93/compiler/frontend/src/org/jetbrains/kotlin/contracts/parsing/PsiContractParserDispatcher.kt#L107) then it doesn't seem to line up.

Reply:
Oh, I didn't think of comparing old behavior, thanks!

It's not lined up, but I guess it shouldn't be 1-to-1. Anyhow, I have to admit my previous conditions were really naive. I rather refer to what `ConeEffectExtractor` is visiting. Note that, visiting doesn't mean extracting effects. So, `unresolvedEffects` could have statements/expressions that are allowed, attempted, yet not resolved ones as well as simply not allowed ones.